ArC fixes for spec compatibility, round 4

Ladicek opened the pull request:
Related to #28558

- ignore static methods annotated `@Inject`, similarly to static fields
- fix dependency injection order
  - this is to pass the AtInject TCK
  - to review the `BeanGenerator` changes, I suggest to hide whitespace changes
- implement CDI 4.0 startup/shutdown events
  - I believe it's best if we let current Quarkus events just extend the CDI 4.0 event classes, so that they are basically the same
- improve bean archive detection
  - treat archives that contain Build Compatible Extensions just like we treat archives that contain Portable Extensions (which is what CDI requires)
  - archives that contain `beans.xml` with `bean-discovery-mode="none"` are no longer treated as bean archives

Ladicek committed:
ArC: ignore static methods annotated @Inject, similarly to static fields

Ladicek committed:
ArC: implement CDI 4.0 startup/shutdown events

Ladicek committed:
ArC: improve bean archive detection

mkouba commented on extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BeanArchiveProcessor.java:
```diff
@@ -157,6 +165,33 @@ private boolean isAdditionalBeanArchive(ApplicationArchive archive,
+    private boolean possiblyBeanArchive(ApplicationArchive archive) {
+        return archive.apply(tree -> {
+            boolean result = true;
+            for (String beansXml : List.of("META-INF/beans.xml", "WEB-INF/beans.xml")) {
```
I don't think you can use WAR dependencies in quarkus but let's make it safe  ;-)

Ladicek replied:
I agree, but the same set of files is inspected by the `isExplicitBeanArchive` method. I would be in favor of only looking at `META-INF/beans.xml`.

mkouba replied:
Do we make this change in this pr?

Ladicek replied:
I suggest not, but I don't really have a good reason other than it seems out of the scope (of improving spec compatibility). I can add a commit to get rid of `WEB-INF/beans.xml` usage if you want.

mkouba replied:
Let's address this in a separate pull request..

mkouba commented on independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/injection/order/InjectionOrderTest.java:
```diff
@@ -0,0 +1,113 @@
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.test.ArcTestContainer;
+
+public class InjectionOrderTest {
```
This test actually does not verify that injection points from the superclass are injected before injection points from the subclass, or am I missing something?

Ladicek replied:
Humm, good point. Yeah, I don't think that is tested -- the test currently only verifies that fields are injected before initializer methods. I think I might have an idea how to test this (though the AtInject TCK is a lot more complete than this test will ever be :-) ).

Ladicek replied:
Unfortunately you can only capture constructor invocation and initializer method invocation, not field injection. The only possibility I'm aware of is checking whether fields are already injected or not in the constructor and especially in the initializer method.

mkouba replied:
> The only possibility I'm aware of is checking whether fields are already injected or not in the constructor and especially in the initializer method.

Yep, makes sense.  If a field is injected then just add a missing "field injection event" in the constructor and initializer... It's not accurate but good enough for the testing purpose I think.

Ladicek replied:
Done.

mkouba commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Injection.java:
```diff
@@ -296,16 +310,78 @@ private static List<AnnotationInstance> getAllInjectionPoints(BeanDeployment bea
 
+    // ---
+
+    // this is not entirely precise (according to the JLS), but it's better than Methods.MethodKey
```
In which sense is it better? Shouldn't we rather improve the `Methods.MethodKey` then?

Ladicek replied:
I was thinking about improving `MethodKey`, but I looked at its usages and wasn't entirely sure it wouldn't break everyone. I'll take a look again.

Ladicek replied:
Done -- improved `Methods.MethodKey` and `Methods.isOverridden()`, removed `Injection.MethodOverrideKey`.

mkouba commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Injection.java:
```diff
@@ -296,16 +310,78 @@ private static List<AnnotationInstance> getAllInjectionPoints(BeanDeployment bea
+        MethodOverrideKey(MethodInfo method) {
+            if (Modifier.isPrivate(method.flags())) {
+                // private methods cannot be overridden
+                this.name = method.declaringClass().name() + "|" + method.name();
```
It wouldn't hurt to explain what the `name` represents in this case...

Ladicek replied:
I'll add a comment to the code, but this also answers the question above -- this is better than `MethodKey`, because it treats visibility correctly (or maybe almost correctly :-); to be fair, I haven't done extensive testing). Only `public` and `protected` methods may be overridden anywhere in the inheritance hierarchy; package-private methods may only be overridden in the same package, and `private` methods may not be overridden at all.

So if a subclass and its superclass both contain a `private` method with the same name and signature, they are 2 different methods and one does _not_ override the other. Same with package-private methods, if they appear in different packages.

Ladicek replied:
Done.

Ladicek committed:
ArC: fix dependency injection order

In case of class beans, dependency injection order is specified
by the AtInject specification as follows:

- injection points on superclasses are injected before injection points
  on subclasses;
- fields are injected before initializer methods.

Since we discover injection points bottom up, this requires reordering
after each class in the hierarchy is scanned for injection points.

The specification also prescribes that overridden initializer methods
are not injected at all. Since we discover injection points bottom up,
this requires remembering already seen methods and skipping methods
that are overridden by those that were previously seen.

This requires proper override detection. The `Methods.isOverridden()`
and `Methods.MethodKey` utils are not enough, because they actually
do _not_ detect overrides. They are used to skip processing a method
in case a method with the same signature has already been processed.
To do detect overridden methods properly, it is important to remember
that:

- private methods cannot be overridden,
- package-private methods can only be overridden in the same package.

For that reason, this commit introduces a dedicated override detection
utility.

mkouba commented:
The CI failures look related:
```
2023-02-20T19:28:06.6408580Z 2023-02-20 19:28:06,610 ERROR [io.sma.rea.mes.provider] (main) SRMSG00212: Unable to initialize mediator: io.quarkus.smallrye.reactivemessaging.signatures.ProcessorSignatureTest$BeanProducingAPublisherOfPayload#process: java.lang.ClassFormatError: Duplicate method name "getEmitter" with signature "()Lio.smallrye.reactive.messaging.annotations.Emitter;" in class file io/quarkus/smallrye/reactivemessaging/signatures/ProcessorSignatureTest_BeanProducingAPublisherOfPayload_ClientProxy
```

Ladicek commented:
Yeah. I'll look.

manovotn commented on extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BeanArchiveProcessor.java:
```diff
@@ -157,6 +165,33 @@ private boolean isAdditionalBeanArchive(ApplicationArchive archive,
+                    // crude but enough
+                    try {
+                        String text = Files.readString(pathVisit.getPath());
+                        if (text.contains("bean-discovery-mode='none'")
```
We could also detect `all` and throw an appropriate exception in order to prevent user errors (because such archive would now still be accepted but treated as `annotated`)

mkouba replied:
I agree that we could at least log a warning. I'm not so sure about an exception because this could break a lot of apps...

Ladicek replied:
Warning is probably a good idea.

manovotn replied:
> because this could break a lot of apps...

Assuming they were accidentally working in the first place as we treat them as `annotated` anyway :-D
Jokes aside, I am fine with warning as well.

Ladicek replied:
Done.

manovotn commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Injection.java:
```diff
@@ -320,7 +320,13 @@ private static List<AnnotationInstance> getAllInjectionPoints(BeanDeployment bea
             if (inject != null) {
-                injectAnnotations.add(inject);
+                if (Modifier.isStatic(method.flags())) {
+                    LOGGER.warn("An initializer method must be non-static - ignoring: "
```
As this is a user error, I'd rather just throw an exception.

mkouba replied:
Makes sense. I wonder why the spec does not treat a static initializer as a definition error :thinking:.

Ladicek replied:
See a few lines above -- this is how we treat `static` fields, I just follow suit :-)

mkouba replied:
I know but it's never late to change your mind! ;-)

Ladicek replied:
It is. If we thrown an exception, ArC wouldn't even be able to process the AtInject TCK classes :-)

manovotn replied:
> See a few lines above -- this is how we treat static fields, I just follow suit :-)

Yes, I noticed that and thought we could change both.

> It is. If we thrown an exception, ArC wouldn't even be able to process the AtInject TCK classes :-)

Hm, I see. In that case it'd be either a material for the strictness switch or we'd need to settle just for warning.

Ladicek replied:
The AtInject specification allows optional `static` injection, and the AtInject TCK has a switch to test this. We [will] set that switch to `false`, obviously, but we still need to be able to process those classes.

Using the strictness switch here would be an option, though not exactly intuitive (I'd expect code that works with strictness enabled to also work without strictness enabled).

manovotn replied:
Not, I think it simply had conditional tests that verify the same functionality even for static variants if enabled. CDI has this disabled.
See https://github.com/eclipse-ee4j/injection-tck/blob/master/src/main/java/org/atinject/tck/Tck.java#L117-L119
And the related class - https://github.com/eclipse-ee4j/injection-tck/blob/master/src/main/java/org/atinject/tck/auto/Convertible.java#L474

manovotn replied:
Ah, sorry for double answer, didn't see Ladislav's comment until I posted mine.

@Ladicek I agree that in this case it's better to keep the warning. However, a comment in the code (near that warning perhaps?) would be really nice :)

Ladicek replied:
Good point. I'll add a comment there.

Ladicek replied:
Done.

manovotn reviewed (commented):
I only have two minor comments.
Many thanks for diving into the AtInject TCKs :)

Ladicek commented:
The test failure shows exactly why I was originally hesitant to change `Methods.MethodKey`: it is often used to skip processing a method in case a method with the same signature has already been processed. This is _not_ a definition of overriding (though it's close). I'll move back to having a separate detection of overrides. The reason why I need to detect overrides properly is of course the AtInject TCK, which verifies that overridden initializer methods are not injected twice, but _not_ overridden methods with the same signature _are_.

Ladicek commented:
I have no idea what the failure means -- the Windows job probably ran a lot later than the other jobs and the cached Maven repo .zip was no longer available? :thinking:

manovotn commented:
> I have no idea what the failure means -- the Windows job probably ran a lot later than the other jobs and the cached Maven repo .zip was no longer available? thinking

I've seen that before but don't know the cause either.
However, it shouldn't be related to the changes here, so I think it'd be safe to merge.
